Restructure NEST 3.0 guides (#1929)
nest/nest-simulator: doc/userdoc/guides/spatial/overview_spatial.rst

.. _topo_changes:

Overview of spatially-structured networks
=========================================

-  All topology functions are now part of ``nest`` and not
   ``nest.topology``
-  You can use the ``Create`` and ``Connect`` functions for spatial  networks, the same as you would for non-spatial
   networks
-  All former topology functions that used to take a layer ID, now take a NodeCollection
-  All former topology functions that used to return node/layer IDs now return a NodeCollection

.. note::

   See the reference section :ref:`topo_ref` in our conversion guide for all changes made to functions

Create spatially distributed nodes
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Creating spatially distributed nodes is now handled with the standard ``nest.Create()`` function.

Reviewer: ```suggestion
Spatially distributed nodes can now be created using the standard ``nest.Create()`` function.
```

> Arguments of node creation have also been changed to make creating
> populations with and without spatial information more unified. To create

Reviewer: ```suggestion
The arguments of this function have been changed to unify the
creation of populations with and without spatial information. To create
```

        nest.spatial.free(nest.random.uniform(),  # Positions generated from a uniform distribution
                          num_dimensions=3,       # in 3D
                          edge_wrap=True)         # with periodic boundary conditions

Note the following

- For positions generated from NEST parameters, the number of neurons
  has to be provided in ``nest.Create()`` with the ``n`` argument.
- The extent is calculated from the positions of the nodes, but can be
  set explicitly.
- If possible, NEST tries to deduce the number of dimensions. But if
  the positions are generated from NEST parameters, and there is no
  extent defined, the number of dimensions has to be provided.

  ::

      spatial_nodes = nest.Create('iaf_psc_alpha', n=5,
                                  positions=nest.spatial.free(nest.random.uniform(),
                                                              num_dimensions=3))

Spatially positioned nodes are no longer subnets, as subnets have been removed, but
are rather NodeCollections with metadata. These NodeCollections behave as normal
NodeCollections with two exceptions:

- They cannot be merged, as concatenating NodeCollections with metadata is

>   not allowed.

Reviewer: ```suggestion
  not supported.
```

Retrieving spatial information
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To retrieve the spatial information from your nodes, spatially structured NodeCollections have
a ``.spatial`` parameter that will retrieve all spatial information as a dictionary.

>>>  spatial_nodes.spatial
     {'center': (0.41717460937798023, 0.3541409997269511, 0.5058779059909284),
      'edge_wrap': False,
      'extent': (0.6786768797785043, 0.4196595948189497, 0.8852582329884171),
      'network_size': 5,
      'positions': ((0.1951471883803606, 0.24431120231747627, 0.5770208276808262),
       (0.34431440755724907, 0.46397079713642597, 0.8201442817226052),
       (0.17783616948872805, 0.4038907829672098, 0.16324878949671984),
       (0.3796140942722559, 0.2643292499706149, 0.848507022485137),
       (0.6565130492672324, 0.38219101540744305, 0.4020354822278023))}

Note that if you have specified your positions as a NEST parameter, NEST will convert that
to a list with lists, and this is what you will get when calling ``.spatial``.

Connect spatially distributed nodes
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Similar to creating nodes with spatial distributions, connecting is now done with the

Reviewer: ```suggestion
Similar to creating nodes with spatial distributions, such nodes are now connected using the
```

standard ``nest.Connect()`` function. Connecting NodeCollections with
spatial data is no different from connecting NodeCollections without
metadata. In a layer-connection context, moving to the standard
``Connect()`` function brings with it some notable changes:

- Convergent and divergent specification of connection is removed, or
  rather renamed. See table below.

  ======================================= ==================================================
  NEST 2.x                                NEST 3.0
  ======================================= ==================================================
  ``convergent``                          ``pairwise_bernoulli`` with ``use_on_source=True``
  ``convergent`` with ``num_connections`` ``fixed_indegree``
  ``divergent``                           ``pairwise_bernoulli``
  ``divergent`` with ``num_connections``  ``fixed_outdegree``
  ======================================= ==================================================

  ``use_on_source`` here refers to whether the mask and connection probability
  should be applied to the source neuron or the target neuron.
  This is only required for ``pairwise_bernoulli``, as ``fixed_indegree``

>   and ``fixed_outdegree`` implicitly states if we are using the source or
>   target nodes.

Reviewer: ```suggestion
  and ``fixed_outdegree`` implicitly only apply to either the source or target nodes.
```

  |                              {'mean': 0.0, 'sigma': 1.0}},       |     nest.Connect(l, l, conn_dict)                                   |
  |                  'delays': {'linear': {'a': 1.5, 'c': 0.}},      |                                                                     |
  |                  'allow_multapses': True,                        |                                                                     |
  |                  'allow_autapses': False}                        |                                                                     |
  |     tp.ConnectLayers(l, l, conn_dict)                            |                                                                     |
  |                                                                  |                                                                     |
  +------------------------------------------------------------------+---------------------------------------------------------------------+

Masks
^^^^^
In NEST 3.0, the mask ``volume`` got removed, as the same mask was already available under the name ``box``.

> The former was only an alias available in NEST for backward compatibility.

Reviewer: ```suggestion
The former was only an alias available for backward compatibility.
```

Retrieving distance information
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
If you have a SynapseCollection with connections from a spatially distributed network, you can retrieve the
*distance* between the source-target pairs by calling ``.distance`` on the SynapseCollection.

  ::

    s_nodes = nest.Create('iaf_psc_alpha', positions=nest.spatial.grid(shape=[3, 1]))
    t_nodes = nest.Create('iaf_psc_alpha', positions=nest.spatial.grid(shape=[1, 3]))
    nest.Connect(s_nodes, t_nodes)

    conns = nest.GetConnections()
    dist = conns.distance

``.distance`` will be a tuple of the same length as your SynapseCollection, where ``dist[indx]`` will be the distance
between the source-target pair at *indx*.

Calling ``.distance`` on a SynapseCollection where either the source or target, or both, are not spatially

> distributed also works, you will receive `nan` whenever one of the nodes is non-spatial.

Reviewer: ```suggestion
distributed also works, and you will receive `nan` whenever one of the nodes is non-spatial.
```
